Fetch & Carry Tutorial (#103)
convince-project/AS2FM: docs/source/tutorials.rst

The output is a JANI file called `main.jani` that will be located in the same folder.

.. _full_tutorial:

> Hands-on In-depth Tutorial Including Verification: Fetch & Carry
> -----------------------------------------------------------------

Reviewer: ```suggestion
Hands-on in-depth tutorial including verification: Fetch & Carry example
--------------------------------------------------------------------------------
```

Author: Why not capitalizing headlines?

Author: Ok, will change it

In this tutorial you will learn within around one hour how a fetch and carry robot scenario can be modeled in SCXML and how linear temporal logic properties can be verified on it. We translate the model of the robot and its environment with AS2FM into JANI for verification with SMC Storm, our statistical model checking tool.
We will observe fulfilled and violated properties. By updating the model with more involved functionality in terms of probabilistic behavior and features in the behavior tree we make in the end sure that all properties hold.

We assume some background in computer science or as a robotics developer but no knowledge about formal methods and model checking is required.

Reference Model: Fetch & Carry Robot
`````````````````````````````````````

> For this tutorial we use the model defined here: `ros_example_w_bt <https://github.com/convince-project/AS2FM/tree/main/test/jani_generator/_test_data/tutorial_fetch_and_carry>`_.

Reviewer: ```suggestion
For this tutorial we use the model defined here: `tutorial_fetch_and_carry <https://github.com/convince-project/AS2FM/tree/main/test/jani_generator/_test_data/tutorial_fetch_and_carry>`_.
```

Author: Thanks, will be fixed in next commit.

What is implemented there is a classical fetch and carry task. A robot should drive to the pantry where food is stored, pick up snacks, drive to the table and place the snacks there. The robot should be done with this task after at most 100 seconds.

The model consists of a `main.xml` file, referencing to the BT `bt.xml` running in the system and the SCXML files modeling the BT plugins for navigating `bt_navigate_action.scxml`, picking `bt_pick_action.scxml`, and placing `bt_place_action.scxml`, as well as the world model `world.scxml`. Finally, there are the properties to check later with SMC Storm in the JANI format in `properties.jani`.

All of those components are summarized and collected in the `main.xml` file.

> * First, some parameters relevant for the model checking in the end to bound the model execution are given. This is the maximum time of execution bound to 100 seconds, the maximal array size allowed is 10, and a flag indicating TODO.

Reviewer: ```suggestion
* First, some parameters configuring generic properties of the system. In this example we bound the maximum execution time to 100 seconds, we configure unbounded arrays to have contain at most 10 elements and we configure the Behavior Tree to be ticked again in case it returns SUCCESS or FAILURE (by default, we tick it for as long as it returns RUNNING, and we stop afterwards).
```

Author: Done with minor corrects to your suggestion.

The third image depicts the behavior of the BT plugin `bt_pick_action.scxml` in a very similar fashion. The action is used to pick a certain item with a given id, stored in `data`. When the BT is ticked it assigns `object_id = data`. When the BT is halted or the action is aborted `tmp_result` is set to false, otherwise it is set to true. Based on that the return status of the tree is then published.

The last image depicts the behavior of the BT plugin `bt_place_action.scxml`. When called, the action just immediately tries to successfully execute, no matter if there is an object in the gripper or not, when the BT is ticked. When the BT is halted or the action is aborted `tmp_result` is set to false, otherwise it is set to true. Based on that the return status of the tree is then published.

As a last step we are having a closer look at the environment model in `world.scxml`.

> * First, it is indicated that the model makes use of dependencies from the `delib_ws_24_interfaces` package, where custom ROS actions are defined. In the last line the ROS topic publisher for the snack type is declared.

Reviewer: Weren't we planning to rename this package into something more generic?

Author: Not sure but would make sense. I will keep that as a todo for later.

This property can be checked by calling SMC Storm on the JANI file generated before with AS2FM. For more details on SMC Storm you can have a look the `SMC Storm repository <https://github.com/convince-project/smc_storm>`_.

Executing SMC Storm on this example works as follows:

.. code-block:: bash

    `smc_storm --model main.jani --properties-names snack_at_table --traces-file traces.csv --show-statistics`

The expected result shown below indicates that the property is fulfilled with probability 1, i.e., the snack is always successfully placed on the table. In this case model checking needed 500 traces to come to that result called with the default confidence and error parameters. The minimal length of a trace generated in those runs was 159 and the maximal length was 237.

.. code-block:: bash

    ============= SMC Results =============
        N. of times target reached:     500
        N. of times no termination:     0
        Tot. n. of tries (samples):     500
        Estimated success prob.:        1
        Min trace length:       159
        Max trace length:       237
    =========================================
    Result: 1

> The changes of the values in the different ROS topics can be inspected by having a look at the log of the traces generated during model checking in `traces.csv`. A tool to inspect them graphically is PlotJuggler. Just run `ros2 run plotjuggler plotjuggler` to open the graphical interface, select `traces.csv` in the file explorer and pull the topic you want to inspect from the topic list into the coordinate system in the main inspection area. For the topic TODO this looks as in the picture below.

Reviewer: One could also load the traces.csv file directly, by appending `-d traces.csv` to the plotjuggler command

Author: Added it in the description. Thanks.

The output is a JANI file called `main.jani` that will be located in the same folder.

> 

Reviewer 2: Avoid more than one empty line.

I think we should add

```
- repo: https://github.com/rstcheck/rstcheck
  rev: main # should be replaced with the current verison
  hooks:
    - id: rstcheck
      additional_dependencies: ['rstcheck[sphinx,toml]']
```

I assume this would find that, but I didn't check.

.. _full_tutorial:

> Hands-on in-depth tutorial including verification: Fetch & Carry
> -----------------------------------------------------------------

Reviewer 2: ```suggestion
Hands-on in-depth tutorial including verification: Fetch & Carry
----------------------------------------------------------------

What you will learn
```````````````````
```

> In this tutorial you will learn within around one hour how a fetch and carry robot scenario can be modeled in SCXML and how linear temporal logic properties can be verified on it. You translate the model of the robot and its environment with AS2FM into JANI for verification with `SMC Storm <https://github.com/convince-project/smc_storm>`_, our statistical model checking tool.

Reviewer 2: ```suggestion
In this tutorial you will learn in about one hour how a robotic example can be expressed in SCXML.
The example models a robot with a simple symbolic fetch and carry task.
You will use linear temporal logic (LTL) properties to verify it is working correctly. 
Then, you will translate the model of the robot and its environment with AS2FM into JANI to run it with `SMC Storm <https://github.com/convince-project/smc_storm>`_, our statistical model checking (SMC) tool.
```

> You will observe fulfilled and violated properties. By updating the model with more involved functionality in terms of probabilistic behavior and additional features in the behavior tree, you make in the end sure that the required properties hold.

Reviewer 2: ```suggestion
With the results from SMC Storm, you will observe that some properties are fulfilled and some are violated.
By updating the model with more complexity in terms of probabilistic behavior and additional features in the behavior tree (BT), you make sure that the required properties hold in the end.
This verifies, that your BT controls the example correctly.
```

> We assume some background in computer science or as a robotics developer but no knowledge about formal methods and model checking is required.

Reviewer 2: ```suggestion
We assume some background in computer science or as a robotics developer but no knowledge about formal methods or model checking is required.
```

Reference Model: Fetch & Carry Robot
`````````````````````````````````````

> For this tutorial we use the model defined here: `tutorial_fetch_and_carry <https://github.com/convince-project/AS2FM/tree/main/test/jani_generator/_test_data/tutorial_fetch_and_carry>`_.

Reviewer 2: I would suggest, the participants first copy the model somewhere. Or we put it in a special tutorial or example folder. Having them work in our test data feels like the wrong place.

> What is implemented there is a classical fetch and carry task. A robot should drive to the pantry where food is stored, pick up snacks, drive to the table and place the snacks there. The robot should be done with this task after at most 100 seconds.

Reviewer 2: ```suggestion
A classical fetch and carry task is implemented there. A robot should drive to the pantry where food is stored, pick up snacks, drive to the table and place the snacks there. The robot should be done with this task after at most 100 seconds.
```

Enhancing the Model with Probabilities
`````````````````````````````````````````
This is a very simple example and behavior of the robot. In real world applications the item which should be brought to another location sometimes slips out of the gripper when trying to pick it. Let's say this happens in 40% of the trials. In addition, navigation fails sometimes, let's say in 30% of the cases. We would like to reflect this scenario by adapting the world model in `world_probabilistic.scxml <https://github.com/convince-project/AS2FM/blob/main/examples/tutorial_fetch_and_carry/world_probabilistic.scxml>`_. From now on we are using `main_probabilistic.scxml <https://github.com/convince-project/AS2FM/blob/main/examples/tutorial_fetch_and_carry/main_probabilistic.scxml>`_, which is the same as `main.scxml <https://github.com/convince-project/AS2FM/blob/main/examples/tutorial_fetch_and_carry/main.scxml>`_ but referencing this modified world model in line 15.

> If you want to try to come up with a solution on your own on how to modify the world model such that it's behavior is probabilistic, try to fill the gaps flagged with `TODO` (sometimes in comments, sometimes directly in the code) in the file `world_probabilistic_gaps.scxml <https://github.com/convince-project/AS2FM/blob/main/examples/tutorial_fetch_and_carry/world_probabilistic_gaps.scxml>`_. Afterwards you can read on here and compare your solution with ours in `world_probabilistic.scxml <https://github.com/convince-project/AS2FM/blob/main/examples/tutorial_fetch_and_carry/world_probabilistic.scxml>`_.

Author: Yes, makes sense. Will accept your change.
